### DISCUSSION TITLE
A pluggable aproach to add any container pocess with k3s executable pocess

### DISCUSSION THREAD

**Discussion by sumantfordev** [k3s-io/k3s#7279] [Q&A]
K3s wraps all important kubernets services like kubeproxy, controlle manager, scheduler etc as a single executable binary. This is great to run k3s in an embedded system/ IOT devices. However there are k3s packaged services that runs as seperate containers like traefic, coreDNS etc. In many cases there are requirements which requires to run some other kubernets services for different use cases. E.g can be to run any other CNI service, or perhaps a log exporter as a daemonset, etc etc. 

My ask is is there a way to integrate any 3rd party service/ container / golang module within the k3s process. so that based on requirement anybody can add an external module/application to be part of either server and/or agent process. 

It doesnot have to a extendable plugin to incorporate a 3rd part app within k3s process but even a simple developer documentation that can be followed can be good enough.

**Comment by brandond** [comment]
> My ask is is there a way to integrate any 3rd party service/ container / golang module within the k3s process. so that based on requirement anybody can add an external module/application to be part of either server and/or agent process.

Not within the main k3s process, no. For it to be executed in the main process, it must be built in to K3s at compile time.

**Reply by sumantfordev** [reply]
which is also OK, to probably clone the k3s code, add perhaps a submodule and inject it in k3s server and/or agent main go program o build it. As long as the k3s code part can be upgraded from this main repo to their forked repos keeping the 3rd party service part almost intact, it may be helpful.
